{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Align reads using bowtie. Filter only reads that aligned and convert the output to a bedfile\n",
    "##### Bedfile will have the positions of every read that aligned to given reference genome (in this case NC_000913.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bowtie -S -m 1 -v 0 indexes/NC_000913.2 netseq_data/clean_Trimmed_wt-mmc.fastq.gz 2> wt_mmc_NET.log|samtools view -b -h -F 4|bedtools bamtobed -i '-' > wt_mmc_NET_aligned_2.bed "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Split alignments by strand and edit the alignment bedfile so that it only keeps the 3' end of each gene\n",
    "##### 'plus_3end.bed' and 'minus_3end.bed' are the outputfiles that now only have the position of the 3' base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "!grep -w '+' wt_mmc_NET_aligned_2.bed|awk '{print $1\"\\t\"$2\"\\t\"($2+1)\"\\t\"\".\"\"\\t\"$5\"\\t\"$6}'> plus_3end_2.bed\n",
    "!grep -w '-' wt_mmc_NET_aligned_2.bed|awk '{print $1\"\\t\"($3-1)\"\\t\"($3)\"\\t\"\".\"\"\\t\"$5\"\\t\"$6}'> minus_3end_2.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import packages needed for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline  \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make a pandas dataframe of each positon of the E. coli genome (index) and the corresponding 3' base counts that are present in that position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_3end=pd.read_table('plus_3end_2.bed',header=None)\n",
    "m_3end=pd.read_table('minus_3end_2.bed',header=None)\n",
    "p_counts=pd.DataFrame(p_3end[2].value_counts()).sort_index().reindex(range(1,4641653),fill_value=0)\n",
    "m_counts=pd.DataFrame(m_3end[2].value_counts()).sort_index().reindex(range(1,4641653),fill_value=0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import bedfile so that gene positions can be determined (in future will also make GFF/GTF file an option)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annot=pd.read_csv(\"NC_000913.2.bed\",sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find gene of interest in the befile and store the coordinates of the gene location (in the future will make it so a list of genes can be given)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene='ivbL'\n",
    "gene_strand=gene_annot[gene_annot[4]==gene].iloc[0][5]\n",
    "gene_pos=(gene_annot[gene_annot[4]==gene].iloc[0][1],gene_annot[gene_annot[4]==gene].iloc[0][2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Only keep counts that reflect transcription happening on the sense strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "if gene_strand == '+':\n",
    "    gene_plot=m_counts.iloc[gene_pos[0]:gene_pos[1],:]\n",
    "if gene_strand == '-':\n",
    "    gene_plot=p_counts.iloc[gene_pos[0]:gene_pos[1],:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot the counts in a barplot (reverse the orientation of the plot if gene is '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([], <a list of 0 Text xticklabel objects>)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3cAAACECAYAAAAz3dI8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAACHFJREFUeJzt3U+IXdUdB/DvT9OuujCSUSSGThZZtLtKUKGbgjT+2cSN0C5qECEbCy10I90EdOOqUKEIoQ2NUFqEFsxCCCEUSheWTEqxtVISGquDoikR2+JKOF3MDTyT+fNm5s2bN+d9PjC89373vHvPve/x7v1y7z1TrbUAAACwt92x2x0AAABg+4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAf27XYH1nPgwIG2uLi4290AAADYFZcvX/53a21hnLYzHe4WFxeztLS0290AAADYFVX1r3HbuiwTAACgA8IdAABAB4Q7AACADsz0PXcAANNy+PC7a067dm1xav0A2Cpn7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB3YMNxV1aGq+n1VvVNVb1fVD4b63VV1oaquDI/7h3pV1ctVdbWq3qqqB0bmdWJof6WqTuzcagEAAMyXcc7cfZ7kR621ryV5OMlzVfX1JM8nudhaO5Lk4vA6SR5PcmT4O5nklWQlDCY5leShJA8mOXUzEAIAALA9G4a71tqHrbU/D8//m+SdJAeTHE9ydmh2NsmTw/PjSV5tK95McldV3Zfk0SQXWms3WmufJLmQ5LGJrg0AAMCc2tQ9d1W1mOQbSf6U5N7W2ofJSgBMcs/Q7GCS90fetjzU1qrfuoyTVbVUVUvXr1/fTPcAAADm1tjhrqq+kuS3SX7YWvvPek1XqbV16l8stHa6tXa0tXZ0YWFh3O4BAADMtbHCXVV9KSvB7lettd8N5Y+Gyy0zPH481JeTHBp5+/1JPlinDgAAwDaNM1pmJflFkndaaz8ZmXQuyc0RL08keX2k/vQwaubDST4dLts8n+RYVe0fBlI5NtQAAADYpn1jtPlmku8l+WtV/WWo/TjJS0leq6pnk7yX5Klh2htJnkhyNclnSZ5Jktbajap6Mcmlod0LrbUbE1kLAACAObdhuGut/TGr3y+XJI+s0r4leW6NeZ1JcmYzHQQAAGBjmxotEwAAgNkk3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA8IdAABAB4Q7AACADgh3AAAAHRDuAAAAOiDcAQAAdEC4AwAA6IBwBwAA0AHhDgAAoAPCHQAAQAeEOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiAcAcAANCBDcNdVZ2pqo+r6m8jtbur6kJVXRke9w/1qqqXq+pqVb1VVQ+MvOfE0P5KVZ3YmdUBAACYT+OcuftlksduqT2f5GJr7UiSi8PrJHk8yZHh72SSV5KVMJjkVJKHkjyY5NTNQAgAAMD2bRjuWmt/SHLjlvLxJGeH52eTPDlSf7WteDPJXVV1X5JHk1xord1orX2S5EJuD4wAAABs0Vbvubu3tfZhkgyP9wz1g0neH2m3PNTWqgMAADABkx5QpVaptXXqt8+g6mRVLVXV0vXr1yfaOQAAgF5tNdx9NFxumeHx46G+nOTQSLv7k3ywTv02rbXTrbWjrbWjCwsLW+weAADAfNlquDuX5OaIlyeSvD5Sf3oYNfPhJJ8Ol22eT3KsqvYPA6kcG2oAAABMwL6NGlTVr5N8K8mBqlrOyqiXLyV5raqeTfJekqeG5m8keSLJ1SSfJXkmSVprN6rqxSSXhnYvtNZuHaQFAACALdow3LXWvrvGpEdWaduSPLfGfM4kObOp3gEAADCWSQ+oAgAAwC4Q7gAAADog3AEAAHRAuAMAAOiAcAcAANAB4Q4AAKADwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQAeEOAACgA/t2uwMAAL04fPjdNaddu7Y4tX4A88mZOwAAgA4IdwAAAB0Q7gAAADog3AEAAHRAuAMAAOiA0TIBOmf0PgCYD87cAQAAdEC4AwAA6IBwBwAA0AH33AEAMLPWum/YPcNwO2fuAAAAOuDMHQAA7DAjFzMNwh3AHueAAWDnuTyUvUC4Y0c42GTW+Y4yD3zPJ88BPjDLph7uquqxJD9NcmeSn7fWXpp2H1ifgwEAYJ449qEXUw13VXVnkp8l+XaS5SSXqupca+3v0+wHMNvsZIF55jcQ2Kppn7l7MMnV1to/k6SqfpPkeBLhjk0bZ+e3V3aQe6Wf49hoXXpa13Fs9xIu22vFZtZ1o3n4jk7eLFyq2NM+oSe2+fTZ5vNt2uHuYJL3R14vJ3loyn3o2nZ2bpM8eJrEPKbx4zStg4FZ2eZbXcakl7Nde+l7vlf09NnPQj+m1c+98ls8ie01K5/9JMzC5zaNz3WW5rFX9qE9fW57ZT8868ebm+3DtMNdrVJrX2hQdTLJyeHl/6rqHzveqzlRq239TUw3j6212ellbLcP45rEcmbhc5uVbT6teczC93wSZmVdZ+X7s90+jNPGPGZvGbM0j+0uYxL92Evbq6d57IVlTMos/F4P07867vymHe6WkxwaeX1/kg9GG7TWTic5Pc1OAQAA7HV3THl5l5IcqarDVfXlJN9Jcm7KfQAAAOjOVM/ctdY+r6rvJzmflX+FcKa19vY0+wAAANCjaq1t3AoAAICZNu3LMgEAANgBwh0AAEAHhDsAAIAOCHcAAAAdEO4AAAA6INwBAAB0QLgDAADogHAHAADQgf8DyXYlFeFOxeIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1080x144 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x=gene_plot.index.tolist()\n",
    "y=gene_plot[2].tolist()\n",
    "if gene_strand=='-':\n",
    "    y=y[::-1]\n",
    "plt.figure(figsize=(15,2))\n",
    "sns.barplot(x=x,y=y,color='blue')\n",
    "sns.despine(top=False,right=False,bottom=False,left=False)\n",
    "plt.xticks([])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
